pithos/plugins/notify.py

```diff
@@ -35,40 +35,38 @@ def on_prepare(self):
             os.environ['GNOTIFICATION_BACKEND'] = 'freedesktop'
 
         self._app = Gio.Application.get_default()
+        self._app_id = self._app.get_application_id()
+        self._fallback_icon = Gio.ThemedIcon.new('audio-x-generic')
         self.prepare_complete()
 
     def on_enable(self):
-        self.song_change_handler = self.window.connect('song-changed', self.send_notification)
-        self.state_change_handler = self.window.connect('user-changed-play-state', self.send_notification)
+        self._song_change_handler = self.window.connect('song-changed', self.send_notification)
 
     def send_notification(self, window, *ignore):
         if window.is_active():
-            return
-
-        song = window.current_song
-        # This matches GNOME-Shell's format
-        notification = Gio.Notification.new(song.artist)
-        notification.set_body(song.title)
-        if song.artUrl:
-            notification.set_icon(Gio.FileIcon.new(Gio.File.new_for_uri(song.artUrl)))
-
-        if window.playing:
-            notification.add_button(_('Pause'), 'app.pause')
+            # GNOME-Shell will auto dismiss notifications
+            # when the window becomes "active" but other DE's may not (KDE for example).
+            # If we're not going to replace a previous notification
+            # we should withdraw said stale previous notification.
+            self._app.withdraw_notification(self._app_id)
         else:
-            notification.add_button(_('Play'), 'app.play')
-        notification.add_button(_('Skip'), 'app.next-song')
+            song = window.current_song
+            # This matches GNOME-Shell's format
+            notification = Gio.Notification.new(song.artist)
+            # Needed for KDE otherwise clicking on our notifications does nothing...
+            notification.set_default_action('app.activate')
+            notification.set_body(song.title)
 
-        if self._last_id != song.trackToken:
-            self._app.withdraw_notification(self._last_id)
+            if song.artUrl:
+                icon = Gio.FileIcon.new(Gio.File.new_for_uri(song.artUrl))
+            else:
+                icon = self._fallback_icon
+            notification.set_icon(icon)
 
-        self._last_id = song.trackToken
-        self._app.send_notification(song.trackToken, notification)
+            notification.add_button(_('Skip'), 'app.next-song')
 
-    def on_disable(self):
-        if self._last_id:
-            self._app.withdraw_notification(self._last_id)
-            self._last_id = ''
-
-        self.window.disconnect(self.song_change_handler)
-        self.window.disconnect(self.state_change_handler)
+            self._app.send_notification(self._app_id, notification)
 
+    def on_disable(self):
+        self._app.withdraw_notification(self._app_id)
+        self.window.disconnect(self._song_change_handler)
```

Comment:
`self._song_change_handler` may be `0`.

Reply:
Sure I'll do a ```if``` to check for that but it never should be 0. ```connect``` always returns a non-zero number and ```on_disable``` is always preceded by ```on_enable```. There is nothing in ```on_prepare``` or  ```on_enable``` that should ever fail.

Reply:
`connect` can return `0` on failure but yes it will probably never fail in our usage. Just call it defensive programming =)